### ISSUE TITLE
SOPFLOW Python Tests re-visit (functionality vs unit tests)

### ISSUE THREAD

**Issue by cameronrutherford** [pnnl/ExaGO#142]
**Issue type**
- [x] Bug

**Relates to**
- [x] SOPFLOW

**Summary**

From [this thread](https://github.com/pnnl/ExaGO/pull/131#discussion_r1586440540) on #131, it is clear that SOPFLOW tests run the entire application many times, however these are only functionality tests. We should also ensure unit tests.

There is a comment in the source code that mentions the inability to do this without calling `sopflow.setup()` or else our destructor segfaults. We should also either:
- [ ] Fix this issue in the destructor, and make sure the correct memory is allocated / freed and the bug is squashed
- [ ] Document this thoroughly, explain why, and make sure users are aware

cc #140 as well since a similar issue was seen there according to @jaelynlitz and @Jayapreethi

**Comment by pelesh** [comment]
@cameronrutherford, has this issue been addressed? Please advise if it needs further attention on can be closed.

**Comment by cameronrutherford** [comment]
I don't think it really needs more attention if users aren't complaining about anything?

Idk I think the issue was that the destructor isn't perfect for SOPFLOW, but that's okay. I don't know if it's worth the effort to debug and squash that unless a use case presents itself. Otherwise it's just for the sake of better tests...

**Comment by pelesh** [comment]
> I don't think it really needs more attention if users aren't complaining about anything?
> 
> Idk I think the issue was that the destructor isn't perfect for SOPFLOW, but that's okay. I don't know if it's worth the effort to debug and squash that unless a use case presents itself. Otherwise it's just for the sake of better tests...

What you describe may be similar to the issue fixed in #190. It might be worthwhile checking it out.

**Comment by cameronrutherford** [comment]
> > I don't think it really needs more attention if users aren't complaining about anything?
> > 
> > Idk I think the issue was that the destructor isn't perfect for SOPFLOW, but that's okay. I don't know if it's worth the effort to debug and squash that unless a use case presents itself. Otherwise it's just for the sake of better tests...
> 
> What you describe may be similar to the issue fixed in #190. It might be worthwhile checking it out.

From the look of the PR it might have fixed it, in which case adding the tests might not be that hard at all.

Just need to make sure those tests are run somewhere in CI...
